### ISSUE TITLE
Matlab README needs updating

### ISSUE THREAD

**Issue by cstackpole** [easybuilders/easybuild-easyconfigs#6295]
Greetings,
I was wanting to install Matlab. I attempted to follow along with the directions. I downloaded the two 2017 iso's, mounted them, rsync'd the contents to a 2017a folder, tar'd it up, and tried doing the install with eb. But it seemed like I was constantly having issues (some of my own doing, some because of Java [shakes fist :-)], some because the README is incomplete). I still don't have it working... I was also really confused about the README saying I needed to update the config file for the license. Huh? There is nothing about the license.

So I went poking around.

Looks like there used to be lines for the license but no longer. Also, it seems that we may not need to jump through unpacking the iso's anymore?? At least that is what I gathered from #408. 

I was so close to pushing up a new version of the README for how I was doing it, but then it failed again...(failing sanity tests and I don't know why). I figured no one wanted updated partial notes that don't work. :-D

Can someone who knows how to install Matlab with EasyBuild update the README for me please? How do I install Matlab with EasyBuild?

Thank you!

**Comment by guest1604** [comment]
Hello @cstackpole,

Here my custom easyconfig file MATLAB-2017b.eb, I hope it can help you. For the source, I have unpacked the ISO files and created a "tar.gz" archive (as stated in README).

`key` parameter is the installation key __required for installing MATLAB__

```
name = 'MATLAB'
version = '2017b'

homepage = 'http://www.mathworks.com/products/matlab'
description = """MATLAB is a high-level language and interactive environment
 that enables you to perform computationally intensive tasks faster than with
 traditional programming languages such as C, C++, and Fortran."""

toolchain = {'name': 'dummy', 'version': ''}

# be sure to copy both DVD content to the SAME directory, 
# including the hidden files, especially .dvd1 and .dvd2 
sources = [SOURCELOWER_TAR_GZ]

dependencies = [('Java', '1.8.0_121')]

key = 'XXXXX-XXXXX-XXXXX-XXXXX-XXXXX-XXXXX-XXXXX-XXXXX-XXXXX-XXXXX-XXXXX-XXXXX-XXXXX'

modluafooter = """
prepend_path("MLM_LICENSE_FILE", "port@licsever1:port@licserver2:port@licserver3")
"""

moduleclass = 'math'
```

**Comment by cstackpole** [comment]
Thanks! I will try this right now!!

**Comment by cstackpole** [comment]
Thanks very much @guest1604 for that. It helped greatly and I found the issue!

A previous internet search uncovered someone who said that I had to set the environment variable MLM_LICENSE_FILE to point to my file (which is a network based file that has two lines (see below since I have flex lm running on a different system). This wasn't working for me. I got errors about binaries not being found.
> SERVER \<IP\> ANY \<port\>
> USE_SERVER

The documentation says to modify and add the following to the eb script, but that part of the code was removed from the files. If you go back in history you can find a section below but it wasn't working for me either. I still got errors about binaries not being found. 
> license_server = 'license.example.com'   
> license_server_port = '00000'   
> key = '00000-00000-00000-00000-00000-00000-00000-00000-00000-00000'   

So I tracked down the issue and this is where I think this turns into a bug report ( @boegel ).

The error is this:
`== FAILED: Installation ended unsuccessfully (build directory: /home/sys/clusteradmin/.local/easybuild/build/MATLAB/2017a/dummy-): build failed (first 300 chars): Sanity check failed: no file of ('bin/matlab',) in /software/eb/software/MATLAB/2017a, no file of ('bin/mcc',) in /software/eb/software/MATLAB/2017a, no file of ('bin/glnxa64/MATLAB',) in /software/eb/software/MATLAB/2017a, no file of ('bin/glnxa64/mcc',) in /software/eb/software/MATLAB/2017a, no fi
== Results of the build can be found in the log file(s) /tmp/eb-Y_jXOu/easybuild-MATLAB-2017a-20180511.102835.OnBhH.log
ERROR: Build of /software/eb/software/EasyBuild/3.6.0/lib/python2.7/site-packages/easybuild_easyconfigs-3.6.0-py2.7.egg/easybuild/easyconfigs/m/MATLAB/MATLAB-2017a.eb failed (err: "build failed (first 300 chars): Sanity check failed: no file of ('bin/matlab',) in /software/eb/software/MATLAB/2017a, no file of ('bin/mcc',) in /software/eb/software/MATLAB/2017a, no file of ('bin/glnxa64/MATLAB',) in /software/eb/software/MATLAB/2017a, no file of ('bin/glnxa64/mcc',) in /software/eb/software/MATLAB/2017a, no fi")
`
But that isn't the error. That is the sanity check failing (so it is doing it's job). The actual error is in the install phase which I missed before because of the other phases printing out success messages! EB should have caught/flagged/failed out at the install step. Here is the errors in the logs at the install phase. 
`(May 11, 2018 10:31:00) Error: You have entered an invalid File Installation Key.
(May 11, 2018 10:31:00) Exiting with status -2
(May 11, 2018 10:31:01) End - Unsuccessful.
Finished
`
As seen, those other methods were not actually picking up the license (or it just doesn't like the Network based license file I gave it). But instead of telling me 'your license is wrong' it gives me false errors about missing binaries...

Note that the only difference between the old documentation and your example is how the license server/port is added. I need to re-investigate this and see if maybe I did something else wrong. I'm wondering why this was yanked. Maybe there is a better way of doing it?

So all is good now, right? I have an install? Well...no. I'm still getting error in the sanity check, but the install says it succeeded and I do actually have an install instead of a blank folder. It seems I still don't have 'bin/glnxa64/mcc' which is causing it to fail. :-/ Researching it now.

Once I have this working, I'm totally documenting this process and pushing it into the README. ;-)

**Comment by guest1604** [comment]
@cstackpole 

I use 
```
modluafooter = """
prepend_path("MLM_LICENSE_FILE", "port@licsever1:port@licserver2:port@licserver3")
"""
```
instead of 
```
license_server = 'license.example.com'
license_server_port = '00000'
```

Because I want to specify **3 license servers,** this is a kind of "workaround" (I don't think that's a good practise). If you have only 1 license server, use `license_server` and `license_server_port ` _(or set MLM_LICENSE_FILE environment variable before installing ?)_.

But **in my case** license servers are not required for installing MATLAB, only the File Activation Key (`key` parameter) is required. License servers are only required for using MATLAB.

`modluafooter` is a parameter that only add lines at the end of the module file, it's not used for installing MATLAB.

> modluafooter | Footer to include in generated module file (Lua syntax) | "" 

ref : https://github.com/easybuilders/easybuild/blob/master/docs/version-specific/easyconfig_parameters.rst

**Comment by cstackpole** [comment]
It looks like I'm getting errors when it tries to compile the 'mcc' binary. According to [mathworks](https://www.mathworks.com/support/sysreq/previous_releases.html) 2017a needs gcc 4.9. The default I had was 4.8. So I just grabbed 4.9.4 from EB and am rebuilding Matlab now. However, already I have a warning. :-/

> WARNING: Found one or more non-allowed loaded (EasyBuild-generated) modules in current environment:
> * GCCcore/4.9.4
> * binutils/2.25-GCCcore-4.9.4

How is yours finishing the install? Are you needing to do anything to pass those sanity checks for 'bin/glnxa64/mcc'?

**Comment by cstackpole** [comment]
Sure. I will check that now.

I just managed my first "successful installation". By removing 'bin/glnxa64/mcc' and 'bin/mcc' from the sanity check...Probably not a great idea in the long run but after 5 hours I needed a win and Matlab seems to be working. I just don't have the Matlab compiler. :D

**Comment by cstackpole** [comment]
In the log file I see this message which why I am hoping a "supported compiler" might fix the issue with the mcc binaries not being created:
```
1. The following products require a supported compiler:

Stateflow 8.9
Simulink Coder 8.12
MATLAB Coder 3.3
Simulink Test 2.2

2. Simulink requires a C compiler for simulation acceleration, model reference, and MATLAB Function Block capabilities. It is recommended that you install a supported compiler on your machine.

3. To accelerate computations with the following products, a supported compiler is required:

SimBiology 5.6
Fixed-Point Designer 5.4
```

**Comment by cstackpole** [comment]
Seems that it failed and I don't know enough about EB yet to work around it.
```
ERROR: Irresolvable dependencies encountered: Java/1.8.0_172-GCC-4.9.4-2.25
```
Rebuilding Java with that GCC doesn't make sense but I don't know how to tell it to ignore that.

**Comment by guest1604** [comment]
Yes you're right, using `dependencies = [('Java', '1.8.0_172'), ('GCC', '4.9.4-2.25')]` is probably a better solution than changing the toolchain because you may need to use MATLAB with others version of GCC. If you change the toolchain you may have issue to swap version of GCC if you use Hierarchical Module Naming Schema.

Just for information, to solve this error `ERROR: Irresolvable dependencies encountered: Java/1.8.0_172-GCC-4.9.4-2.25` you can use this in your easyconfig :

`dependencies = [('Java', '1.8.0_172', '', ('dummy', ''))]`
or
`dependencies = [('Java', '1.8.0_172', '', True)]`

Reference : https://github.com/easybuilders/easybuild/blob/master/docs/Writing_easyconfig_files.rst (Specifying dependencies using dummy toolchain)

**Comment by cstackpole** [comment]
Even with passing that GCC version, still no go on the mcc binary. So either:
* I'm doing something wrong, which I really don't know what the issue is.
* Everyone else just comments out the sanity check too. :-P

The reason why I think that I'm doing something wrong is because I have these files in my installs in the other environment (where I have MANY versions of Matlab). I'm just at a loss as to why they are not being installed via EB.

**Comment by cstackpole** [comment]
BTW, here is my rough steps for recreating. 
```
# First download both dvd iso's from Mathworks.
$ mount -o loop,ro R2017a_glnxa64_dvd1.iso /mnt/
$ rsync -avHlP /mnt/ R2017a/
$ umount /mnt
$ mount -o loop,ro R2017a_glnxa64_dvd2.iso /mnt/
$ rsync -avHlP /mnt/ R2017a/
$ umount /mnt
$ tar -zcvf /my/easybuild/download/path/matlab-2017a.tar.gz R2017a
$ eb MATLAB-2017a.eb -r
```
And just for completeness here's my MATLAB-2017a.eb file (edited of course). The install license key I am using (don't know if it matters) is for MDCS so it should actually install every toolbox it has:
```
name = 'MATLAB'
version = '2017a'

homepage = 'http://www.mathworks.com/products/matlab'
description = """MATLAB is a high-level language and interactive environment
 that enables you to perform computationally intensive tasks faster than with
 traditional programming languages such as C, C++, and Fortran."""

toolchain = {'name': 'dummy', 'version': ''}

# be sure to copy both DVD content to the SAME directory,
# including the hidden files, especially .dvd1 and .dvd2
sources = [SOURCELOWER_TAR_GZ]

dependencies = [('Java', '1.8.0_172')]

key = '<edited>'
license_server = '<edited>'
license_server_port = '<edited>'

moduleclass = 'math'
```
And finally, I had to edit the easy block file [matlab.py](https://github.com/easybuilders/easybuild-easyblocks/blob/master/easybuild/easyblocks/m/matlab.py) which I really don't think is the right thing to do but it's the only way it completes for me:
```
Line 141
-            'files': ["bin/matlab", "bin/mcc", "bin/glnxa64/MATLAB", "bin/glnxa64/mcc",
+            'files': ["bin/matlab", "bin/glnxa64/MATLAB",
```

**Comment by guest1604** [comment]
Maybe, you should test MATLAB manually installation on this host and check if mcc binaries are installed in order to know if this issue is related to EasyBuild or not.

**Comment by cstackpole** [comment]
Good point. I'm calling it day (one can only do so many Matlab installs in a day...:-) for now but I will test this next. Thanks for the suggestions and help today!

**Comment by boegel** [comment]
@cstackpole MATLAB easyblock is enhanced to catch the `invalid File Installation Key` error in https://github.com/easybuilders/easybuild-easyblocks/pull/1423 .

W.r.t. the missing `mcc`: can you check whether you see `Installing Product: MATLAB Compiler` in the output of the installation command? Are there any particular `product.*` lines uncommented in the input file that is passed to the installer? If so, `product.MATLAB_Compiler` must be one of them in order to get `mcc` installed (if all are commented out, all 'products' supported by your license should be installed, which should also include `mcc`).

**Comment by cstackpole** [comment]
So - after FAR too many installs of Matlab on a Saturday, I know why I'm having issues with EB installing Matlab. I'm just not sure what to do about it.

When I use the graphical installer (which is what I've done in the past), I use the "Log in to Mathworks" option. My user base brings their own license, but list me as an Admin so I can update the license files without having to pester them. I select the MDCS license because it is what installs the most toolboxes. I think the GUI log-in is picking up that there are other licenses with other products because when I do the install this way I also have installed the mcc binaries which come from the Compiler toolbox (which some of my users have). However, if I pass the license key only (via gui or via command line) I do NOT get the Compiler toolbox nor the mcc binaries.

Guess which toolbox is on the list of [ineligible for MDCS](https://www.mathworks.com/products/ineligible_programs.html)?  
Yup, the Compiler.  
And what toolbox ISN'T installed when passing the license key to EB?  
Yup, the Compiler.  
And again, what provides the mcc binaries?  
Yup, the Compiler.  

I don't know how to get Matlab to install EVERYTHING from the command line without having a single license for EVERY toolbox. Nor do I know how to pass it multiple keys/do-whatever-it-does-when-it-installs-with-mathworks-log-in.

My current theory here is that I can do the bulk of the installs with EB going forward with the MDCS license. Then I can do another install manually with a license that has the compiler and point it at the same directory (I've done that before, Matlab has in the past picked up the old install and only installed the difference).

As for the toolboxes that it claims I don't have a supported compiler for (Stateflow, Simulink Coder, MATLAB Coder, Simulink Test, SimBiology, Fixed-Point Designer) well it turns out I can't find any license from my users with these in them and it doesn't look like they were installed in past versions either. So, oh well! I guess that was just a wild herring (though I am kinda curious what is required for them to be installed - it might be interesting to add that info to EB). In fact, of the licenses that my users have that are NOT a part of MDCS it is only the Compiler that is missing!

Sigh. I really hate software license keys. They never make things easier. But that is another rant. :-P

Let me work out a few things and then in the near future I will put together a README update for the Matlab install that hopefully will help the next admin that comes along avoid most of the same pain I just went through. No one should have to do this many installs of Matlab. :-D

Thanks for the help!

**Comment by boegel** [comment]
Thanks a lot for the details @cstackpole!

That sure sounds like a frustrating endavour, ugh...

Maybe you should reach out to MATLAB support on this?

**Comment by cstackpole** [comment]
A big update.
1) If you ask nicely you can have Mathworks add a "All Product File Installation Key" flag to a license.
* This does NOT grant you a license for all the products, but it let's you install the bulk.
* "All Product File Installation" is a lie. It isn't all products. There are exceptions. I know for a fact that MDCS is not installed with a "All Product File Installation". You still have to do double installs to get MDCS (the second on top of the first with different license keys). Their response was that it is an "all client toolboxes install" which, to me, seems dishonest. [Why not just say that in the title? But that is not soapbox for this group. :-) ]. The point being that you can't expect all the server tools to be installed with the "All Product File Installation Key" so those of us that also need those tools will have multiple installs on top of each other to do.

This leads me to my first two questions:
Q1) I know the -f flag forces a complete reinstall, but is there a "just install on top of it" option? I don't see one but I could also be looking for something so specific that I'm missing something obvious.
Q2) How feasible would it be to pass in an array and have EB do multiple installs with multiple keys? 

2) The messages about "The following products require a supported compiler"? Yeah, completely 100% red herring. They are pointless and are printed out regardless. Doesn't matter what compiler is installed or will be installed, the actual install process never bothers to check. If you have those toolboxes, it prints those messages. So it really doesn't matter if EB adds in / recommends the use of a GCC compiler or not.

Q3) Is it worth putting into the README file? Maybe something about "matching the compiler and Matlab version is left to the admin and out of scope for this eb script"? I dunno. On one hand, this would have to be handled post Matlab install when the user starts the toolbox and does the config (at least from my understanding of it) so it is likely that any admin who knows about / uses these toolboxes already knows that they need to configure GCC but on the other hand someone who is expecting EB to install Matlab and also configure GCC to work with these toolboxes might be confused when it isn't done. Maybe it is worth a mention for the small amount of EB users who need these toolboxes? Thoughts?

3) I also confirmed that the mcc files only come from the compiler toolbox. I'm not sure what the ratio is of people who use EB and have the compiler vs don't have the compiler. 

Q4) Should the sanity check for the compiler be default? Or should we add that into the documentation for over-riding the sanity check variables?

Q5) Should EB sanity check toolboxes based on what was installed? That could get messy. My thoughts would be just to do sanity checks on the base Matlab as it is the only thing that is pretty much guaranteed to be installed every time and it is unlikely that it would be installed correctly yet the rest of the install failed (regardless of which or how many other toolboxes are selected for install).

Q6) This isn't so much relevant to the above. What are the thoughts about maybe providing a "template file" in the README that would host the license key, sanity overrides, ect that can be passed into the EB Matlab script? That way instead of having EB users modify every script they can just have a reference file in their EB directory. Seems like a good idea, but I don't know what it would take to make that happen yet.

Thanks!

**Comment by cstackpole** [comment]
I wrote up my directions and the pull request above works. However, those directions do not yet account for the questions I posted above.

**Comment by lpouillo** [comment]
Hello,

by adding the dependency on GCC-4.9.3 and providing both licence_server and key, I succesfully install my new Matlab 2017a version, with mcc and mex.

Default Centos 4.8.5 compilers was not working.

Thanks for giving the information that help me to solve my installation

system : Centos 7.4
easybuild : 3.6.1

Best regards
